File: R/nrmlepln.R
#' Full information maximum likelihood and bivariate composite likelihood estimation for
#' polytomous logit-normit (graded logistic) model
#' @aliases nrmlerasch nrbcpln
#' @description Full information maximum likelihood and bivariate composite likelihood estimation for
#' polytomous logit-normit and Rasch models, via Newton Raphson iterations.
#' @usage
#' nrmlepln(x, ncat, nitem=NULL, alphas=NULL, betas=NULL, abound=c(-10,10),
#'          bbound=c(-1,10), nq=48, mxiter=200, m2=TRUE, iprint=FALSE)
#' nrmlerasch(x, ncat, nitem=NULL, alphas=NULL, abound=c(-10,10),
#'          bbound=c(-1,10), nq=48, mxiter=200, m2=TRUE, iprint=FALSE)
#' nrbcpln(x, ncat, nitem=NULL, alphas=NULL, betas=NULL, abound=c(-10,10),
#'          bbound=c(-1,10), nq=48, mxiter=200, se=TRUE, iprint=FALSE)
#' @param x A data matrix. Data can be in one of two formats: 1) raw data
#'  where the number of rows corresponds to an individual's response and
#'  each column represents an item, and 2) a matrix of dimensions
#'  \code{nrec}\eqn{\times}{ X }(\code{nitem}+1) where each row corresponds to a
#'  response pattern and the last column is the frequency of that response
#'  pattern. A data matrix of the second type requires input for \code{nitem}
#'  and \code{nrec}.
#' @param ncat Number of ordinal categories for each item, coded as
#'  0,\dots,(\code{ncat}-1). Currently supported are items that have the same number of categories.
#' @param nitem Number of items. If omitted, it is assumed that \code{x} contains
#'  a data matrix of the first type (raw data) and the number of
#'  columns in \code{x} will be selected as the number of items.
#' @param alphas A vector of length \code{nitem}\eqn{\times}{ X }(\code{ncat}-1)
#'  corresponding to starting values for the (decreasing) cutpoints
#'  for the items. If omitted, these will be computed from the function \code{\link{startalphas}}.
#' @param betas A vector of length \code{nitem} corresponding to starting values for the
#'  beta vectors of slopes. If omitted, these will be computed from the function
#' \code{\link{startbetas}}. For the polytomous logit-normit, there is one slope
#'  for each item; for the Rasch model, there is a common slope beta for
#'  all of the items.
#' @param abound Vector of length 2 that sets upper and lower bounds on parameter estimation for
#'  alphas. Currently experimental; changing defaults it not recommended. Estimation problems
#'  are more likely solved by changing starting values.
#' @param bbound Vector of length 2 that sets upper and lower bounds on parameter estimation for
#'  betas. Currently experimental; changing defaults it not recommended. Estimation problems
#'  are more likely solved by changing starting values.
#' @param nq Number of quadrature points to use during estimation. This argument is currently experimental. It is recommended to use the default of 48.
#' @param mxiter Maximum number of iterations for estimation.
#' @param se Logical. If \code{TRUE}, calculates standard errors for the bivariate composite
#'  likelihood method.
#' @param m2 Logical. If \code{TRUE}, computes goodness-of-fit statistics
#'  from Maydeu-Olivares and Joe (2005, 2006; i.e., \eqn{M_2}).
#' @param iprint Logical. Enables debugging / diagnostic information from C code that conducts
#'  estimation.
#'  
#' @details
#'  Estimation of graded logistic models is performed under the following parameterization:
#'  \deqn{Pr(y_i = k_i| \eta) = \left\{
#'  \begin{array}{ll}
#'  1-\Psi (\alpha_{i,k} + \beta_i \eta) & \mbox{if  } k_i = 0\\
#'  \Psi (\alpha_{i,k} + \beta_i \eta) - \Psi (\alpha_{i,k+1} + \beta_i \eta) & \mbox{if  } 0 < k_i < m-1\\
#'  \Psi (\alpha_{i,k+1} + \beta_i \eta) & \mbox{if  } k_i = m-1
#'  \end{array} \right. }{
#'  Pr(y_i = k_i| \eta) = {
#'  1-\Psi (\alpha_i,k + \beta_i*\eta)  if k_i = 0, 
#'  \Psi (\alpha_i,k + \beta_i*\eta) - \Psi (\alpha_i,k+1 + \beta_i*\eta)  if 0 < k_i < m-1,  
#'  \Psi (\alpha_i,k+1 + \beta_i*\eta)  if k_i = m-1}.
#'  }
#'  Where the items are \eqn{y_i, i = 1, \dots, n}, and response categories are \eqn{k=0, \dots, m-1}. \eqn{\eta} is the latent trait, \eqn{\Psi} is the logistic distribution function, \eqn{\alpha} is an intercept (cutpoint) parameter, and \eqn{\beta} is a slope parameter. When the number of categories for the items is 2, this reduces to the 2PL parameterization:
#'  \deqn{Pr(y_i = 1| \eta) = \Psi (\alpha_1 + \beta_i \eta)}
#'  Both \code{nrmlepln} and \code{nrbcpln} perform estimation under these parameterizations, via Newton Raphson iterations, using full information maximum likelihood (\code{nrmlepln}) and bivariate composite likelihood (\code{nrbcpln}). See Maydeu-Olivares and Joe (2005, 2006) for more information on bivariate composite likelihood estimation (see also Varin, Reid, and Firth, 2011). Under \code{nrmlerasch} a common \eqn{\beta} parameter is estimated for all items.
#'  
#' @return A list containing the following slots.
#' @slot alphas A vector of parameter estimates for alphas. Length is
#'  \code{nitem}\eqn{\times}{ X }(\code{ncat}-1). Estimates are in order by item, e.g., all alphas
#'  for item 1, followed by all alphas for item 2, and so on.
#' @slot betas A vector of parameter estimates for betas. Length is \code{nitem}.
#' @slot nllk Negative (composite) log-likelihood for polytomous 
#'  logit-normit (or Rasch) model.
#' @slot conv Integer indicating whether estimation converged. Currently only returned for
#'  composite likelihood estimation.
#' @slot sealphas A vector of standard errors for the alpha estimates.
#' @slot sebetas A vector of standard errors for the beta estimates.
#' @slot invhes Inverse Hessian matrix for the MLE estimates.
#' @slot vcov Asymptotic covariance matrix for the composite likelihood estimates.
#' @slot teststat Value of \eqn{M_2}.
#' @slot df Degrees of freedom for \eqn{M_2}.
#' @slot pval P-value for \eqn{M_2}.
#'  
#' @references
#'  Bartholomew, D., Knott, M., and Moustaki, I. (2011). \emph{Latent Variable 
#'  Models and Factor Analysis: A Unified Approach}, 3rd Edition. Wiley. 
#'  
#'  Maydeu-Olivares, A., and Joe, H. (2005). Limited and full information estimation
#'  and goodness-of-fit testing in \eqn{2^n} contingency tables: A unified framework.
#'  \emph{Journal of the American Statistical Association, 100}, 1009-1020.
#'  
#'  Maydeu-Olivares, A., and Joe, H. (2006). Limited information and goodness-of-fit
#'  testing in multidimensional contingency tables. \emph{Psychometrika, 71},
#'  713-732.
#'  
#'  Varin, C., Reid, N. and Firth, D. (2011). An overview of composite likelihood
#'  methods. \emph{Statistica Sinica, 21}, 5-42.
#'  
#' @author Carl F. Falk \email{cffalk@gmail.com}, Harry Joe
#' 
#' @examples
#'  ### Matrix of response patterns and frequencies
#'  data(item5fr)
#'  
#'  ## ML estimation
#'  nrmleplnout<-nrmlepln(item5fr, ncat=3, nitem=5)
#'  print(nrmleplnout)
#'  
#'  ## BCL estimation
#'  nrbcplnout<-nrbcpln(item5fr, ncat=3, nitem=5)
#'  print(nrbcplnout)
#'  
#'  ## ML Rasch estimation
#'  nrmleraschout<-nrmlerasch(item5fr, ncat=3, nitem=5)
#'  print(nrmleraschout)
#'  
#'  \donttest{
#'  ### Raw data
#'  data(item9cat5)
#'  
#'  ## ML estimation
#'  nrmleplnout<-nrmlepln(item9cat5, ncat=5)
#'  print(nrmleplnout)
#'  
#'  ## BCL estimation
#'  nrbcplnout<-nrbcpln(item9cat5, ncat=5, se=FALSE)
#'  print(nrbcplnout)
#'  
#'  ## ML Rasch estimation
#'  nrmleraschout<-nrmlerasch(item9cat5, ncat=5)
#'  print(nrmleraschout)
#'  }
#'
#' @seealso  
#'  \code{\link{startalphas}}
#'  \code{\link{startbetas}}
#'
#' @importFrom stats pchisq
#' @useDynLib pln, .registration=TRUE
#' @export
nrmlepln <- function (x, ncat, nitem=NULL, alphas=NULL, betas=NULL, abound=c(-10,10),
    bbound=c(-1,10), nq=48, mxiter=200, m2=TRUE, iprint=FALSE) {

  myInput<-check.input(x, ncat, nitem, nq, mxiter, iprint)

  ## get starting values if not present already
  if(!check.alphas(alphas, myInput$nitem, myInput$ncat)){
    alphas<-startpln.func(myInput$nitem, myInput$ncat, myInput$nrec, myInput$myX)$alphas
  }

  ## prep betas
  if(!check.betas(betas, myInput$nitem)){
    betas<-startbetas.func(myInput$myX)
  }
        
  ## check bounds
  abound<-check.bounds(alphas, abound)
  bbound<-check.bounds(betas, bbound)

  nrmleplnout <- nrmlepln.func(myInput$nitem, myInput$ncat, myInput$nrec, myInput$myX, alphas,
        betas, abound, bbound, myInput$nq, myInput$mxiter, myInput$iprint)
  alphas<-nrmleplnout$mlePlnOut[1:((myInput$ncat-1)*myInput$nitem)]
  betas<-nrmleplnout$mlePlnOut[((myInput$ncat-1)*myInput$nitem+1):(myInput$ncat*myInput$nitem)]
  
  V<-matrix(nrmleplnout$invHesOut, nrow=myInput$nitem*myInput$ncat,ncol=myInput$nitem*myInput$ncat)
  seVec<-nrmleplnout$seVecOut
    
  sealphas<-seVec[1:((myInput$ncat-1)*myInput$nitem)]
  sebetas<-seVec[((myInput$ncat-1)*myInput$nitem+1):(myInput$ncat*myInput$nitem)]
    
    
  out<-list(alphas=alphas,betas=betas,nllk=nrmleplnout$nllkOut,sealphas=sealphas,
    sebetas=sebetas,invhes=V)

  if(m2){
    m2out<-m2pln.func(myInput$nitem, myInput$ncat, myInput$nrec, myInput$myX, alphas, betas,
      myInput$nq, myInput$iprint)
    pval<-pchisq(m2out$m2stat, m2out$df, lower.tail=FALSE)
      ##out<-append(out, list(samplemout=m2out$samplem, m2stat=m2out$m2stat, df=m2out$df, pval=pval))      
    out<-append(out, list(teststat=m2out$m2stat, df=m2out$df, pval=pval))
  }

  return(out)
}

nrmlepln.func <- function(nitem, ncat, nrec, myX, alphas, betas, abound, bbound, nq, mxiter,
    iprint){
  
  ## prep return variables
  nllkOut<-0
  iconv<-0
  np<-ncat*nitem
  mlePlnOut<-rep(0,np)
  seVecOut<-rep(0,np)
  invHesOut<-matrix(0,nrow=np,ncol=np)

  ## TO DO: add PACKAGE argument here
  out <- .C("Rnrmlepln",
            as.integer(nitem), as.integer(ncat), as.integer(nrec), as.double(myX),
            as.double(alphas), as.double(betas), as.double(abound), as.double(bbound),
            nllkOut=as.double(nllkOut), mlePlnOut=as.double(mlePlnOut),
            seVecOut=as.double(seVecOut), invHesOut=as.double(invHesOut),
            as.integer(nq), as.integer(mxiter), iconv=as.integer(iconv),
            as.integer(iprint))
  return(out)
  
}

File: src/nrmlepln.c
#ifndef R
#include <malloc.h>
#else
#include <R.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* log-likelihood for polytomous logit-normit model
   using Newton Raphson with computed Hessian */

#define SQRTPI 1.772453850905516

#ifdef DATA
/* gcc -DDATA -o nrmlepln nrmlepln.c nrmin.o amatrix.o ssgauher.c polyder2.c -lm */

int main(int argc, char *argv[])
{ double *param,*lb,*ub,fnval,**invhes,bdd;
  double **dmatrix(int, int);
  double ***gmatrix(int, int, int);
  void gauher(double *, double *, int);
  void plnnllk(int, double *, double *, double **, int,
      int, int, int, int, double *, double *, double **, double *,
      double ***, double ***, double ***);
  int j,np,n2,i,ip,jp,nn;
  int mxiter,iprint,iconv;
  double **alp,*b,tmp;
  FILE *in;
  void nrmin(int,double *, void (*)(int, double *, double *, double **, int,
      int, int, int, int, double *, double *, double **, double *,
      double ***, double ***, double ***), 
      double *lb, double *ub, int, double,int, double bdd, 
      int *, double *, double **,
      int, int, int, int, double *, double *, double **, double *,
      double ***, double ***, double ***);
  double *x,*w;
  int nq,n,nrec,m;
  double **dat,*fr;
  double ***g, ***g1, ***g2;

  setbuf(stdout,NULL);
  if(argc==1) { printf("%s parfile [nq] < datafile", argv[0]); exit(0); }
  if(argc>=3) nq=atoi(argv[2]); else nq=48;
  if(nq>48) nq=48;
  printf("\nnq=%d\n",nq);
  x=(double *) malloc((nq+1) * sizeof(double));
  w=(double *) malloc((nq+1) * sizeof(double));
  gauher(x,w,nq);
  for (j=1;j<=nq;j++) x[j]*=M_SQRT2;
  for (j=1;j<=nq;j++) w[j]/=SQRTPI;
  /*for (j=1;j<=nq;j++) printf("%3d %14.6e %14.6e\n",j,x[j],w[j]);*/

  in=fopen(argv[1],"r");
  mxiter=20;
  /*mxiter=1;*/
  fscanf(in, "%d %d %d", &n,&m,&nrec); /* number of items and categories */
  for(i=1,n2=1;i<=n;i++) n2*=m;
  printf("\nn=%d, #categ=%d, nrec=%d\n", n,m,nrec);
  np=m*n;

  ub=(double *) malloc(np * sizeof(double));
  lb=(double *) malloc(np * sizeof(double));
  for(ip=0;ip<np;ip++) { lb[ip]=-10; ub[ip]=10; }
  /* cannot put lower bound of 0 on betas unless
     all items oriented positively */
  for(ip=np-n;ip<np;ip++) { lb[ip]=-1; }
    /* TO DO: set bounds as an argument passed from R*/  
    /*for(ip=np-n;ip<np;ip++) { lb[ip]=0; }*/
  ip=0;

  /* alp[i][1],...,alp[i][m-1] should be in decreasing order */
  alp=dmatrix(n+1,m);
  b=(double *) malloc((n+1) * sizeof(double));

  param=(double *) malloc(np * sizeof(double));
  for(i=1;i<=n;i++)
  { for(j=1;j<m;j++) 
    { fscanf(in, "%lf", &tmp); alp[i][j]=tmp;
      param[ip]=tmp; ip++; }
  }

  for(i=1;i<=n;i++) 
  { fscanf(in, "%lf", &tmp); b[i]=tmp; param[ip]=tmp; ip++; }
  fclose(in);
 
  dat=dmatrix(nrec,n);
  fr=(double *) malloc((nrec) * sizeof(double));
  for(i=0,nn=0;i<nrec;i++)
  { for(j=0;j<n;j++) scanf("%lf", &dat[i][j]);
    scanf("%lf", &fr[i]); nn+=fr[i];
    /*for(j=0;j<n;j++) printf("%d ", dat[i][j]); 
      printf(": %d\n", fr[i]); */
  }
  /*for(j=0;j<n;j++) printf("%f ", dat[nrec-1][j]); */
  printf("sample size=%d\n", nn);

  invhes=dmatrix(np,np);
  iprint=1;
  bdd=5.;  /* bound the difference between iterations */
  g=gmatrix(n+1,m+1,nq+1);
  g1=gmatrix(n+1,m+1,nq+1);
  g2=gmatrix(n+1,m+1,nq+1);

  nrmin(np,param,plnnllk,lb,ub,mxiter,1.e-4,iprint,bdd,&iconv,&fnval,invhes,
    n, m, nq, nrec, x, w, dat, fr, g, g1, g2);
  printf("negative log-likelihood = %f\n", fnval);
  printf("Parameter estimates alp[i][j], j=1..m-1, i=1..n; b[i]\n");
  for(ip=0;ip<np;ip++) 
  { printf("%f ", param[ip]); if(ip%8==7) printf("\n"); }
  printf("\n");
  printf("SEs alp[i][j], j=1..m-1, i=1..n; b[i]\n");
  for(ip=0;ip<np;ip++) 
  { printf("%f ", sqrt(invhes[ip][ip])); if(ip%8==7) printf("\n"); }
  printf("\n");
  /* can comment out for testing */
  printf("\ninverse Hessian\n");
  for(ip=0;ip<np;ip++) 
  { for(jp=0;jp<np;jp++) printf("%.10f ", invhes[ip][jp]); printf("\n"); }
  

  free(invhes[0]); free(invhes);
  free(alp[0]); free(alp);
  free(b);
  free(dat[0]); free(dat);
  free(fr);
  free(param);
  free(g[0][0]); free(g[0]); free(g);
  free(g1[0][0]); free(g1[0]); free(g1);
  free(g2[0][0]); free(g2[0]); free(g2); 
  free(w); free(x);
  free(ub); free(lb);
  exit(0);
}
#endif

void plnnllk(int np, double *param, double *nllk, double **dd, int iprint,
    int n, int m, int nq, int nrec, double *x, double *w, double **dat, double *fr,
    double ***g, double ***g1, double ***g2)
{ int i,j,ir,ip,iq,jp,nfr;
  void plgngh2(int, int, int[], double *, double *, double **,
    int, double *, double *,
    double ***, double ***, double ***);
  double **dmatrix(int, int);
  double tem,tem2,xx,bb,pr;
  int *kk;
  double *der,**hes;
  double **alp,*b;
  /*extern double *x,*w,**dat,*fr;*/
  /*extern int nq,n,nrec,m;*/

  alp=dmatrix(n+1,m);
#ifndef R
  b=(double *) malloc((n+1) * sizeof(double));
#else
  b=(double *) calloc((n+1), sizeof(double));
#endif
  ip=0;
  for(i=1;i<=n;i++)
  { for(j=1;j<m;j++) 
    { alp[i][j]=param[ip]; ip++; }
  }
  for(i=1;i<=n;i++) { b[i]=param[ip]; ip++; }

  for(ip=0;ip<np;ip++)
  { for(jp=0;jp<=np;jp++) dd[ip][jp]=0.; }

  /* set up global arrays when parameters are fixed in an iteration */
  for(i=1;i<=n;i++)
  { bb=b[i];
    for(iq=1;iq<=nq;iq++)
    { g[i][0][iq]=1.; g[i][m][iq]=0.;
      g1[i][0][iq]=0.; g1[i][m][iq]=0.;
      g2[i][0][iq]=0.; g2[i][m][iq]=0.;
      xx=x[iq];
      for(j=1;j<m;j++) 
      { tem=1./(1.+exp(-alp[i][j]-bb*xx)); 
        g[i][j][iq]=tem;
        tem2=tem*(1.-tem);
        g1[i][j][iq]=tem2;
        g2[i][j][iq]=tem2*(1.-2.*tem);
      }
    }
  }
#ifndef R
  kk=(int *) malloc((n+1) * sizeof(int));
  der=(double *) malloc((np+1) * sizeof(double));
#else
  kk=(int *) calloc((n+1), sizeof(int));
  der=(double *) calloc((np+1), sizeof(double));
#endif
  hes=dmatrix(np+1,np+1);

  for(ir=0,*nllk=0.;ir<nrec;ir++)
  { for(i=1;i<=n;i++) kk[i]=dat[ir][i-1];
    nfr=fr[ir];
    if(nfr==0) continue;
    /*pr=plgngh(n,m,alp,b,n,ii,kk,nq);*/
    /* simplify as ii is fixed here and r=n */
    /*pr=plgngh(n,m,alp,b,kk);*/
    plgngh2(n,m,kk,&pr,der,hes,
      nq,x,w,g,g1,g2);
    /* don't update hessian if pr<=0 */
    if(pr<=0.) { pr=1.e-10; /*printf("ir=%d, pr=%f\n", ir,pr);*/ }
    else
    { /* Hessian matrix plus gradient, set up for nrmin in matrix dd[][] */
      for(ip=1;ip<=np;ip++)
      { dd[ip-1][np]-= nfr*der[ip]/pr;
        for(jp=1;jp<=np;jp++)
        { tem=hes[ip][jp]-der[ip]*der[jp]/pr;
          dd[ip-1][jp-1]-= nfr*tem/pr;
        }
      }
    }  
    *nllk -= nfr*log(pr); 
    /*printf("ir=%d, pr=%e, nllk=%f\n", ir,pr,*nllk);*/
  }

  /* for testing print out gradient */
#ifndef R
  if(iprint==1)
  { printf("\ngradient\n");
    for(ip=0;ip<np;ip++) printf("%f ", dd[ip][np]); 
    printf("\n\n");
  }
#else
  if(iprint==1)
  { Rprintf("\ngradient\n");
    for(ip=0;ip<np;ip++) Rprintf("%f ", dd[ip][np]); 
    Rprintf("\n\n");
  }    
#endif
    
#ifndef R
  free(kk);
  free(der); free(hes[0]); free(hes);
  free(alp[0]); free(alp); free(b);
#else
  free(kk);
  free(der); free(hes[0]); free(hes);
  free(alp[0]); free(alp); free(b);
#endif
}

#ifdef R

void Rnrmlepln( int *nitem, int *ncateg, int *nrec, double *dataset,
               double *alphas, double *bvec, double *abound, double *bbound,
               double *nllkout, double *mleplnout, double *sevecout,
               double *invhesout, int *nq, int *mxiter,int *iconv, int *iprint) 
{
   double *param,*lb,*ub,fnval,**invhes,bdd; 
   double **dmatrix(int, int); 
   double ***gmatrix(int, int, int); 
   void gauher(double *, double *, int); 
   void plnnllk(int, double *, double *, double **, int, 
       int, int, int, int, double *, double *, double **, double *, 
       double ***, double ***, double ***); 
   int j,np,n2,i,ip,nn,jp; 
   double **alp,*b; 
   void nrmin(int,double *, void (*)(int, double *, double *, double   **, int, 
       int, int, int, int, double *, double *, double **, double *, 
       double ***, double ***, double ***),  
       double *lb, double *ub, int, double,int, double bdd,  
       int *, double *, double **, 
       int, int, int, int, double *, double *, double **, double *, 
       double ***, double ***, double ***); 
   double *x,*w; 
   int n;
   double **dat,*fr; 
   double ***g, ***g1, ***g2; 

   n=*nitem;

   /*nq=48; */
   x=(double *) calloc((*nq+1), sizeof(double)); 
   w=(double *) calloc((*nq+1), sizeof(double)); 
   gauher(x,w,*nq); 

   for (j=1;j<=*nq;j++) x[j]*=M_SQRT2; 
   for (j=1;j<=*nq;j++) w[j]/=SQRTPI;
    
   /*for (j=1;j<=nq;j++) Rprintf("%3d %14.6e %14.6e\n",j,x[j],w[j]);*/
  
   /*printf("%d %d %d %f\n", n, m, nrec,(*mxGetPr(prhs[1])));*/
 
   /* convert to matrices in C (i.e. row/column transpose) */
   dat=dmatrix(*nrec,*nitem); 
   fr=(double *) calloc((*nrec), sizeof(double)); 
   /* nn = total of fr[] */
   for(i=0,nn=0;i<*nrec;i++) 
   { for(j=0;j<*nitem;j++) dat[i][j] = *(dataset + (j*(*nrec)+i));  
     fr[i]=*(dataset + (i+(*nitem)*(*nrec))); nn+=fr[i];
     /*if(i<10) printf("%f %f\n", dat[i][n-1], fr[i]);*/
   } 

   for(i=1,n2=1;i<=*nitem;i++) n2*=(*ncateg);
   if (*iprint==1)
   {
     Rprintf("\nn=%d, #categ=%d, nrec=%d\n", *nitem,*ncateg,*nrec);
   }
   np=(*ncateg)*(*nitem); 

   lb=(double *) calloc(np, sizeof(double)); 
   ub=(double *) calloc(np, sizeof(double)); 
   for(ip=0;ip<np;ip++) { lb[ip]=*(abound+0); ub[ip]=*(abound+1); } 

  /* cannot put lower bound of 0 on betas unless
     all items oriented positively */
   for(ip=np-n;ip<np;ip++) { lb[ip]=*(bbound+0); ub[ip]=*(bbound+1);}
   ip=0; 

   /* alp[i][1],...,alp[i][m-1] should be in decreasing order */

   alp=dmatrix((*nitem)+1,(*ncateg)); 
   param=(double *) calloc(np, sizeof(double)); 
   for(i=0;i<(*nitem);i++)
   { for(j=0;j<((*ncateg)-1);j++)  
     { alp[i+1][j+1]=*(alphas + ((*ncateg)-1)*i+j); param[ip]=alp[i+1][j+1]; ip++;
       /*if(*iprint==1)
       {
         Rprintf("alpha(%d,%d)=%f\n",i,j,alp[i+1][j+1]);
       }*/
     } 
   } 

   b=(double *) calloc(((*nitem)+1), sizeof(double));  
   for(i=0;i<(*nitem);i++)   
   { param[ip]=*(bvec + i); b[i]=*(bvec + i); ip++;  
     /*if(*iprint==1)
     {
       Rprintf("beta(%d)=%f\n",i,b[i]);
     }*/
   }  

   /*for(j=1;j<=n;j++) printf("%d ", dat[nrec][j]);*/

  /*printf("First row: %d %d %d %d %d %d\n", dat[0][0], dat[0][1],
      dat[0][2], dat[0][3], dat[0][4], fr[0]);
    printf("\nn=%d, #categ=%d, nrec=%d\n", n,m,nrec);
    printf("nn=%d\n", nn);
  */

   invhes=dmatrix(np,np);  
   bdd=5.;  /* bound the difference between iterations */

   g=gmatrix((*nitem)+1,(*ncateg)+1,*nq+1);  
   g1=gmatrix((*nitem)+1,(*ncateg)+1,*nq+1);  
   g2=gmatrix((*nitem)+1,(*ncateg)+1,*nq+1);  

   nrmin(np,param,plnnllk,lb,ub,*mxiter,1.e-4,*iprint,bdd,iconv,&fnval,invhes, 
     *nitem, *ncateg, *nq, *nrec, x, w, dat, fr, g, g1, g2);  

  /* below used for testing only */
  /* printf("negative log-likelihood = %f\n", fnval);
  printf("Parameter estimates alp[i][j], j=1..m-1, i=1..n; b[i]\n");
  for(ip=0;ip<np;ip++) 
  { printf("%f ", param[ip]); if(ip%8==7) printf("\n"); }
  printf("\n");
  printf("SEs alp[i][j], j=1..m-1, i=1..n; b[i]\n");

  for(ip=0;ip<np;ip++) 
  { printf("%f ", sqrt(invhes[ip][ip])); if(ip%8==7) printf("\n"); }
  printf("\n"); */
 
  /*printf("inverse Hessian\n");
  for(ip=0;ip<np;ip++) 
  { for(jp=0;jp<np;jp++) printf("%.10f ", invhes[ip][jp]); printf("\n"); }
  */

  /* Creation of output */

   *nllkout = fnval;

   for(ip=0;ip<np;ip++) *(mleplnout + ip)=param[ip];  

   for(ip=0;ip<np;ip++) *(sevecout + ip)=sqrt(invhes[ip][ip]);  

   for(ip=0;ip<np;ip++)   
   { for(jp=0;jp<np;jp++) *(invhesout + (jp+ip*np))=invhes[ip][jp];}  

   free(invhes[0]); free(invhes);  
   free(alp[0]); free(alp);  
   free(b);  
   free(dat[0]); free(dat);  
   free(fr); free(param);  
   free(w); free(x);  
   free(ub); free(lb);  
   free(g[0][0]);  
   free(g[0]); free(g);  
   free(g1[0][0]);  
   free(g1[0]); free(g1);  
   free(g2[0][0]);  
   free(g2[0]); free(g2);  
}

#endif
